doc/source/nfloat.rst

```diff
@@ -0,0 +1,270 @@
+with precision in small fixed multiples of the word size
+(64, 128, 192, 256, ... bits on a 64-bit machine). The exponent range is
+close to a full word.
+A number with `n`-limb precision is stored using `n+2` limbs as follows:
+
+    +---------------+
+    | exponent limb |
+    +---------------+
+    |   sign limb   |
+    +---------------+
+    |  mantissa[0]  |
+    +---------------+
+    |      ...      |
+    +---------------+
+    | mantissa[n-1] |
+    +---------------+
+
+This type has the advantage that floating-point numbers with the
+same precision can be packed together tightly in vectors
+and created on the stack without heap allocation.
+The precision of an ``nfloat`` context object and its elements cannot
+be changed; to switch precision, one must convert to a different
+context object. For higher precision than supported by ``nfloat``
+and for calculations that require fine-grained precision adjustments,
+one should use :type:`arf_t` instead.
+
+The focus is on fast calculation, not bitwise-defined results.
+Atomic operations typically give slightly worse than correct rounding,
+e.g. with 1-2 ulp error. The rounding is not guaranteed to be identical
+on 64-bit and 32-bit machines.
+Planned features include:
+
+* Directed rounding modes
+* Support for special values (partially implemented)
+* Optional (but slower) IEEE 754-like semantics
+* Complex and ball types
+
+This module is designed to use the :ref:`generics <gr>` interface.
+As such, the domain is represented by a :type:`gr_ctx_t` context object,
+methods return status flags (``GR_SUCCESS``, ``GR_UNABLE``, ``GR_DOMAIN``),
+and one can use generic structures such as :type:`gr_poly_t` for
+polynomials and :type:`gr_mat_t` for matrices.
+
+Types, macros and constants
+-------------------------------------------------------------------------------
+
+.. macro :: NFLOAT_MIN_LIMBS
+            NFLOAT_MAX_LIMBS
+
+    The number of limbs `n` permitted as precision. The current
+    limits are are `1 \le n \le 33` on a 64-bit machine and
+    `1 \le n \le 66` on a 32-bit machine, permitting precision
+    up to 2112 bits. The upper limit exists so that elements and
+    temporary buffers are safe to allocate on the stack and so that
+    simple operations like swapping are not too expensive.
+
+.. type:: nfloat_ptr
+          nfloat_srcptr
+
+    Pointer to an ``nfloat`` element or vector of elements of any
+    precision. Since this is a void type, one must cast to the correct
+    size before doing pointer arithmetic, e.g. via the ``GR_ENTRY``
+    macro.
+
+.. type:: nfloat64_struct
+          nfloat128_struct
+          nfloat192_struct
+          nfloat256_struct
+          nfloat384_struct
+          nfloat512_struct
+          nfloat1024_struct
+          nfloat2048_struct
+          nfloat64_t
+          nfloat128_t
+          nfloat192_t
+          nfloat256_t
+          nfloat384_t
+          nfloat512_t
+          nfloat1024_t
+          nfloat2048_t
```

Comment:
I suppose the numbers will be half of this for 32-bit systems? Or is this module only available for 64-bit computers?

Reply:
My bad. I suppose this just assumes 64-bit system.

Reply:
No, those types should work the same on both 32-bit and 64-bit systems (except for minor rounding differences). They just have twice as many limbs on 32-bit.